### PR TITLE
add command version - #89

### PR DESCRIPTION
**Pull request by SergeyMi37**
Get register  and client versions
zpm: USER>ver

 zpm 0.2.2
 zpm-registry not found
 http://192.168.20.157:52663/registry/ 0.0.2
zpm: USER>

--Get modules versions in all namespaces
zpm: USER>ver -list

  %SYS>
isc-apptools-lockdown 1.0.5
zpm 0.2.2

  USER>
isc-apptools-admin 1.0.11
isc-tar 0.1.2
swagger-ui 1.0.0

### CODE REVIEW

File: src/cls/_ZPM/PackageManager.cls [hunk 1]
```diff
@@ -280,6 +293,138 @@ ClassMethod ShellInternal(pCommand As %String, Output pException As %Exception.A
...
+		if $DATA(list("zpm")) {
+				write !," zpm "_$LG(list("zpm"),1)
+		}
+		set ns=""
+		for { set ns=$order(namespace(ns)) q:ns=""
+			kill list
+			set $namespace=ns
+			do ..GetListModules("zpm",.list)
+			if $DATA(list("zpm-registry")) {
+				write !,name_">zpm-registry "_$LG(list("zpm-registry"),1)
```
Comment (nsolov) [inline src/cls/_ZPM/PackageManager.cls:348]: undefined variable "name"

File: src/cls/_ZPM/PackageManager.cls [hunk 2]
```diff
@@ -280,6 +293,138 @@ ClassMethod ShellInternal(pCommand As %String, Output pException As %Exception.A
...
+				set Found=name
+			}
+		}
+		set $namespace="%SYS"
+		if '$d(Found) write !," zpm-registry not found"
```
Comment (nsolov) [inline src/cls/_ZPM/PackageManager.cls:353]: Let's write a more accurate "Locally installed zpm-registry not found"

File: src/cls/_ZPM/PackageManager.cls [hunk 3]
```diff
@@ -280,6 +293,138 @@ ClassMethod ShellInternal(pCommand As %String, Output pException As %Exception.A
...
+				$$$ThrowOnError(tSC)
+				w " "_o.version
+			}
+		}
+	}
+	q $$$OK
+}
+
+/// write $System.Status.GetErrorText(##class(%ZPM.PackageManager).GetJsonURL("https://pm.community.intersystems.com/",.o))
+ClassMethod GetJsonURL(url As %String, ByRef obj) As %Status
```
Comment (nsolov) [inline src/cls/_ZPM/PackageManager.cls:388]: I suggest moving GetJsonURL Method to class %ZPM.PackageManager.Client.REST.PackageManagerClient, where all methods to get data from the remote registry, is located.

File: src/cls/_ZPM/PackageManager.cls (hunk 1 above)
Comment (SergeyMi37) [reply src/cls/_ZPM/PackageManager.cls]: fixed

File: src/cls/_ZPM/PackageManager.cls (hunk 2 above)
Comment (SergeyMi37) [reply src/cls/_ZPM/PackageManager.cls]: ok

File: src/cls/_ZPM/PackageManager.cls (hunk 3 above)
Comment (SergeyMi37) [reply src/cls/_ZPM/PackageManager.cls]: I moving GetJsonURL Method

File: src/cls/_ZPM/PackageManager.cls [hunk 4]
```diff
@@ -280,6 +289,88 @@ ClassMethod ShellInternal(pCommand As %String, Output pException As %Exception.A
...
+			write !," zpm "_$LG(list("zpm"),1)
+	}
+	set ns=""
+	for { set ns=$order(namespace(ns)) q:ns=""
+		kill list
+		set $namespace=ns
+		do ..GetListModules("zpm",.list)
+		if $DATA(list("zpm-registry")) {
+			write !,ns_">zpm-registry "_$LG(list("zpm-registry"),1)
+			set Found=name
```
Comment (nsolov) [inline src/cls/_ZPM/PackageManager.cls:336]: must be 
`set Found=ns`

Comment (SergeyMi37) [reply src/cls/_ZPM/PackageManager.cls:336]: Yes, Sorry

File: src/cls/_ZPM/PackageManager.cls [hunk 5]
```diff
@@ -280,6 +289,88 @@ ClassMethod ShellInternal(pCommand As %String, Output pException As %Exception.A
...
+		set url=$lg(repo(repo),2)
+		if $GET(url)'="" {
+			w !," ",url
+			// Get version registry
+			Set tClient = ##class(%ZPM.PackageManager.Client.REST.PackageManagerClient).%New()
+			set tSC= tClient.GetJsonURL(url,.o)
+			$$$ThrowOnError(tSC)
+			w " "_o.version
+		}
+	}
```
Comment (daimor) [inline src/cls/_ZPM/PackageManager.cls:370]:
```suggestion
	Set tRes = ##class(%ZPM.PackageManager.Client.RemoteServerDefinition).ExtentFunc()
	While tRes.%Next(.tSC) {
		$$$ThrowOnError(tSC)

		#dim tRepository As %ZPM.PackageManager.Client.RemoteServerDefinition
		Set tRepository = ##class(%ZPM.PackageManager.Client.RemoteServerDefinition).%OpenId(tRes.ID,,.tSC)
		$$$ThrowOnError(tSC)
		Set tService = tRepository.GetPackageService()
		Set tInfo = tService.GetInfo()
		Write !,tRepository.URL," - ",tInfo.version
	}
```

File: src/cls/_ZPM/PackageManager/Client/REST/PackageManagerClient.cls [hunk 6]
```diff
@@ -7,6 +7,45 @@ Property Username As %String;
...
+	Return:(req.HttpResponse.StatusCode = 404) $$$ERROR($$$GeneralError,"Repository doesn't exist OR you don't have access")
+	Return:((req.HttpResponse.StatusCode = 403) && (req.HttpResponse.GetHeader("X-RATELIMIT-REMAINING")=0)) $$$ERROR($$$GeneralError,"API rate limit exceeded. Try logging in.")
+ 	Return:(req.HttpResponse.StatusCode '= 200) $$$ERROR($$$GeneralError,"Received " _ req.HttpResponse.StatusCode _ " status, expected 200")
+
+ 	#dim obj As %ZEN.proxyObject
+	Set st = ##class(%ZEN.Auxiliary.jsonProvider).%ConvertJSONToObject(req.HttpResponse.Data,,.obj,1)
+	Return:$$$ISERR(st) st
+
+ 	Return $$$OK
+}
```
Comment (daimor) [inline src/cls/_ZPM/PackageManager/Client/REST/PackageManagerClient.cls:47]:
```suggestion
Method GetInfo(url As %String) As %DynamicObject
{
 	Set tRequest = ..GetHttpRequest()
 	Set tSC = tRequest.Get()

  If $$$ISOK(tSC), tRequest.HttpResponse.StatusCode=200 {
    Set tRes = {}.%FromJSON(tRequest.HttpResponse.Data)
    Return tRes
  }

	$$$ThrowStatus($$$ERROR($$$GeneralError,"Registry server not available."))
}
```

**Comment by SergeyMi37** [comment]
Dmitry, I checked. Works correctly

**Pull request merged**